File: scrap/sites/stethonet.py
import dateparser
import hashlib
import logging
import re


class Stethonet:
    def __init__(self, browser):
        self.browser = browser

    def fetch_articles(self):
        self.browser.visit(
            "http://www.stethonet.org/annoncephp/recherchedep.php?categorie=0"
        )
        self.browser.select("dep", "69")
        self.browser.find_by_name("Submit").first.click()
        articles = list(
            map(
                lambda a: self.parse_article(a, "Cession/Association/Colaboration"),
                self.browser.find_by_tag("ul"),
            )
        )
        logging.info("Stethonet: %d articles", len(articles))
        self.browser.visit(
            "http://www.stethonet.org/annoncephp/recherchedep.php?categorie=20"
        )
        self.browser.select("dep", "69")
        self.browser.find_by_name("Submit").first.click()
        articles += list(
            map(
                lambda a: self.parse_article(a, "Offre de remplacement"),
                self.browser.find_by_tag("ul"),
            ),
        )
        logging.info("Stethonet: %d articles", len(articles))
        return articles

    def parse_article(self, article, type):
        properties = {"source": self.browser.url, "type": type}
        regex = r"^N°(\d+) - (\d{2}\/\d{2}\/\d{2}) - (.*?)\nCode postal : ([0-9 ]+)$"
        matches = re.match(regex, article.text, re.MULTILINE | re.DOTALL)
        if matches is None:
            properties["description"] = article.text
            properties["id"] = hashlib.md5(
                (properties["description"]).encode()
            ).hexdigest()
        else:
            properties["id"] = matches.group(1)
            # 22/11/19
            properties["date"] = dateparser.parse(
                matches.group(2),
                languages=["fr"],
                locales=["fr-FR"],
            )
            properties["description"] = matches.group(3)
            properties["location"] = matches.group(4)
        logging.debug("Stethonet: article n°%s", properties["id"])
        return properties

File: web/lib/scrap/stethonet.ts
import moment from 'moment';
import { ElementHandle, Browser, Page } from 'puppeteer';
import { logger } from '../logger';
import { Article } from './../../model/article';
import { buildRegexp, md5 } from './utils';

export async function scrapStethonet(browser: Browser) {
    logger.debug("Scrap stethonet");
    return (await Promise.all([
        fetchArticles(
            browser,
            "http://www.stethonet.org/annoncephp/recherchedep.php?categorie=0",
            "Cession/Association/Collaboration"
        ),
        fetchArticles(
            browser,
            "http://www.stethonet.org/annoncephp/recherchedep.php?categorie=20",
            "Offre de remplacement"
        )
    ])).flat();
};

const fetchArticles = async (browser: Browser, url: string, type: string) => {
    const page = await browser.newPage();
    await page.goto(url);
    await page.select("[name=dep]", "69");
    await page.click("[name=Submit]");
    const articles = await page.$$("ul").mapSequential(parseArticle(page, type));
    logger.debug({
        url,
        type,
        articles: articles.length
    }, `Fetched Stethonet articles`)
    return articles;
}

const parseArticle = (page: Page, type: string) => async (articleElement: ElementHandle<Element>): Promise<Article> => {
    const baseProperties = {
        source: page.url(),
        type
    }
    const regex = buildRegexp(
        /^N°(?<id>\d+)/,
        /\s*-\s*/,
        /(?<date>\d{2}\/\d{2}\/\d{2})/,
        /\s*-\s*/,
        /(?<description>.*?)/,
        /Code postal : (?<location>[0-9]+)$/
    );
    const text: string = await page.evaluate(el => el.textContent, articleElement);
    const [match] = [...text.matchAll(regex)];
    if (match?.groups) {
        const { groups } = match;
        const id = groups['id'];
        const date = moment(groups['date'], 'DD/MM/YY').toISOString();
        const description = groups['description'];
        const location = groups['location'];
        const article = {
            ...baseProperties,
            id,
            title: `Annonce n°${id}`,
            date,
            description,
            rawLocation: location
        };
        logger.debug(article, "Stethonet article")
        return article;
    } else {
        logger.error({
            ...baseProperties,
            text
        }, 'Failed to parse Stethonet article');
        return {
            ...baseProperties,
            description: text,
            id: md5(text),
            title: `Annonce non reconnue`,
            date: moment().toISOString(),
        }
    }
}